### PR TITLE
Introduce TypeBranch and TypeMerge instructions

### PR DESCRIPTION
**Pull request by manuranga**
Part of #883

### CODE REVIEW

**Comment by manuranga** [comment]
I ran into a bit of an issue.
let's consider the following case:
```ballerina
public function main() {
    int|string x = 1;
    boolean|float y = true;
    int|float z = 0.2;
    if x is int && y is float && z is int {
        int _ = x;
        float _ = y;
        int _ = z;
    }
}
``` 
In the current draft PR I am generating code as follows:
before

I tried to get rid of the extra BBs to get following code:
after

I am able to generate the above but it has backward branches. Looks like I need to create the `type_branches` in the opposite order I create the BBs.

**Comment by manuranga** [comment]
Just for the record above was a misunderstanding, the backwards-ness is determined by the semantics (what needs to be removed to form DAG?) not by the BB label order.

**Comment by manuranga** [comment]
I found another issue. Consider
```ballerina
public function main() {
    int|string|boolean|float|decimal x = 1;
    if x is int|string|boolean && x is int|string && x is int {
        int _ = x;
        return;
    }
    string|boolean|float|decimal _ = x;
}
```
I currently create:



In order to get the correct result, I ought to generate a single merge per `if`, instead of generating pair-wise. Functionality is not affected.

**Comment by jclark** [comment]
What is the graph you should produce?

**Comment by manuranga** [comment]
Same graph, but with a single type merge.
`x.5 = type_merge [bb0 %x.2] [bb1 %x.4] [bb3 %x.7]`

File: compiler/modules/bir/bir.bal [hunk 1]
```diff
@@ -298,6 +297,8 @@ public enum InsnName {
     INSN_TYPE_TEST,
```
Comment (jclark) [inline compiler/modules/bir/bir.bal:296]: Need to remove COND_NARROW

File: compiler/modules/bir/bir.bal [hunk 2]
```diff
@@ -628,47 +629,24 @@ public type TypeTestInsn readonly & record {|
...
-    InsnRef insn;
-    // Result of the insn
-    boolean result;
+public type TypeMergeInsn readonly & record {|
+    *InsnBase;
+    INSN_TYPE_MERGE name = INSN_TYPE_MERGE;
+    NarrowRegister result;
+    Register[] operands;
+    // Must be predecessors.length() > 1
+    Label[] predecessors;
```
Comment (jclark) [inline compiler/modules/bir/bir.bal:648]: Need to say predecessors.length() == operands.length()

File: compiler/modules/front/stmt.bal [hunk 3]
```diff
@@ -24,15 +24,27 @@ type Assignment readonly & record {|
...
+
+type CondTarget record {|
     bir:BasicBlock block;
-    bir:Register|boolean result;
-    Narrowing[] ifTrue = [];
-    Narrowing[] ifFalse = [];
+    TypeMergeSrc? src = ();
+|};
+
+type TypeMergeSrc record {|
```
Comment (jclark) [inline compiler/modules/front/stmt.bal]: TypeMergeOrigin

File: compiler/modules/front/stmt.bal [hunk 4]
```diff
@@ -272,40 +287,39 @@ function codeGenOnPanic(StmtContext cx, Position pos) {
...
-// If the scope doesn't complete normally, will return empty assignments and non-empty narrowings.
-// Caller should ignore those narrowings and assume it narrows all local variables to NEVER.
-function codeGenScope(StmtContext cx, bir:BasicBlock bb, Environment initialEnv, s:StmtBlock|s:IfElseStmt scope, Narrowing[] initialNarrowing = []) returns CodeGenError|StmtEffect {
+function codeGenScopeForCondTarget(StmtContext cx, CondTarget target, Environment initialEnv, s:StmtBlock|s:IfElseStmt scope) returns CodeGenError|StmtEffect {
+    var { bindings } = initTarget(cx.exprContext(initialEnv), target, scope.startPos);
+    return codeGenScope(cx, target.block, initialEnv, scope, bindings);
+}
+
+// If the scope doesn't complete normally, will return empty assignments and bindings.
+function codeGenScope(StmtContext cx, bir:BasicBlock bb, Environment initialEnv, s:StmtBlock|s:IfElseStmt scope, Binding? initialBinding = ()) returns CodeGenError|StmtEffect {
```
Comment (jclark) [inline compiler/modules/front/stmt.bal]: Say what a scope is

File: compiler/modules/front/stmt.bal [hunk 5]
```diff
@@ -272,40 +287,39 @@ function codeGenOnPanic(StmtContext cx, Position pos) {
...
     Environment env = environmentCopy(initialEnv);
+    if initialBinding != () {
```
Comment (jclark) [inline compiler/modules/front/stmt.bal]: This looks weird

File: compiler/modules/front/stmt.bal [hunk 6]
```diff
@@ -349,26 +363,66 @@ function codeGenStmt(StmtContext cx, bir:BasicBlock? curBlock, Environment env,
 
 function applyEffect(Environment env, StmtEffect effect) {
```
Comment (jclark) [inline compiler/modules/front/stmt.bal]: applyStmtEffect

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front/stmt.bal [hunk 7]
```diff
@@ -24,15 +24,27 @@ type Assignment readonly & record {|
...
 type CondExprEffect record {|
+    // Expression is unconditionally true or false
+    boolean uncond;
+    CondTarget target;
+|}|record {|
+    CondTarget trueTarget;
+    CondTarget falseTarget;
+|};
+
+type CondTarget record {|
```
Comment (jclark) [inline compiler/modules/front/stmt.bal]: TypeMerger

File: compiler/modules/front/expr.bal [hunk 8]
```diff
@@ -302,6 +294,27 @@ function codeGenExpr(ExprContext cx, bir:BasicBlock bb, t:SemType? expected, s:E
...
+    else {
+        bir:AssignTmpRegister result = cx.createAssignTmpRegister(t:BOOLEAN, pos);
+        bir:BasicBlock contBlock = cx.createBasicBlock();
+        assignAndJoin(cx, true, trueMerger.dest, result, contBlock.label, pos);
+        assignAndJoin(cx, false, falseMerger.dest, result, contBlock.label, pos);
+        return { block: contBlock, result };
+    }
+}
+
+function assignAndJoin(ExprContext cx, boolean value, bir:BasicBlock block, bir:AssignTmpRegister result, bir:Label contBlock, Position pos) {
```
Comment (jclark) [inline compiler/modules/front/expr.bal]: Name of this is not OK. Name needs to make it clear that it is mutating `block`.

`block` should be before `value`.

It's creating an assign and a branch insn. The fact that you are using it for joins is not part of the semantic of the function. So something like addAssignAndBranch.

**Review by jclark — CHANGES_REQUESTED** [review]
I will look again after this is rebased with BindingChain change.

File: compiler/modules/nback/scaffold.bal [hunk 9]
```diff
@@ -190,7 +190,14 @@ class Scaffold {
...
         foreach int i in 0 ..< reprs.length() {
             bir:Register register = code.registers[i];
-            self.addresses.push(builder.alloca(reprs[i].llvm, (), register.name));
+            string? name;
+            if register is bir:NarrowRegister {
+                name = register.underlying.name;
+            }
+            else {
+                name = register.name;
+            }
```
Comment (manuranga) [inline compiler/modules/nback/scaffold.bal:199]: This should be `unnarrow(register).name`

**Pull request merged**